### ISSUE TITLE
linux-aarch64 CI tests are being timed out resulting in test failures

### ISSUE THREAD

**Issue by nikhil-arm** [pytorch/pytorch#136192]
> NOTE: Remember to label this issue with "`ci: sev`"

Blocks merging of : https://github.com/pytorch/pytorch/pull/134124 and https://github.com/pytorch/pytorch/pull/135857

## Current Status
*Status could be: preemptive, ongoing, mitigated, closed. Also tell people if they need to take action to fix it (i.e. rebase)*.

## Error looks like
https://github.com/pytorch/pytorch/actions/workflows/linux-aarch64.yml 
https://github.com/pytorch/pytorch/actions/runs/10894440149/job/30233134789

## Incident timeline (all times pacific)
Linux-aarch64 time out is observed since last week and its still on going

## User impact
Blocks testing / merging

## Root cause
*What was the root cause of this issue?*

## Mitigation
*How did we mitigate the issue?*

## Prevention/followups
*How do we prevent issues like this in the future?*

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @seemethere @malfet @pytorch/pytorch-dev-infra @snadampal @milpuz01

**Comment by malfet** [comment]
Remove ci:sev as it’s not really affecting developers except ones who are relying on Linux Aarch64

**Comment by malfet** [comment]
To triage what cause a regression, kicked tests on trunk for
- Sep 10th 2024: https://hud.pytorch.org/pytorch/pytorch/commit/5134ba7458d3073fbe64af9892fcbd674f088604
- Sep 2nd 2024: https://hud.pytorch.org/pytorch/pytorch/commit/982e27e532a54fb7db341cb2537335cb2f026d86
- Aug 30th 2024: https://hud.pytorch.org/pytorch/pytorch/commit/92f282ca520377f054d3c06a873ac093f797145b

**Comment by malfet** [comment]
Please note, that sharding will not work as aarch64 testing setup is very non-standard, i.e. every shard will run the same:
https://github.com/pytorch/pytorch/blob/8e5bb356e083c9dec78927be92baa5a59a3e4d83/.ci/pytorch/test.sh#L1386

**Comment by malfet** [comment]
Attempting to actually make sharding work in https://github.com/pytorch/pytorch/pull/136208

**Comment by abhishek-iitmadras** [comment]
> [Edit] Enabled sharding in #136208, which solved the timeouts, but now testing finishes a bit too soon, which makes we wonder if some tests are missing as results of the breaking them down across the shards. Or if one of the tests alters PyTorch global state.
> 
> See https://github.com/pytorch/pytorch/actions/runs/10923489079/job/30325027660 as an example of a successful run, with 4 shards roughly taking 50 min...

Hi @malfet 
Can we cross-check this by comparing the pre-sharding and post-sharding test counts using `pytest --collect-only` ?

**Comment by malfet** [comment]
> Can we cross-check this by comparing the pre-sharding and post-sharding test counts using `pytest --collect-only` ?

@abhishek-iitmadras that would be nice, please do not hesitate to propose a PR that collects the info and report the results here. Though getting numbers pre-sharding might be challenging, as workflow timed out

**Comment by janeyx99** [comment]
As a meta point--is there a way in the future to detect when a run_test command in CI fails to include shard options? It was probably unexpected that increasing the number of shards did nothing.

A crude solution would be to use lint (either -k or shard options are specified) or to have the script verify when IS_CI is True.

**Comment by clee2000** [comment]
AI: make script to check if sharding works + doesn't skip any tests
AI: math doesn't check out after sharding: timed out on 1 shard, but now takes 40 minutes per shard w/ 4 shards

**Comment by clee2000** [comment]
> AI for @clee2000 : math doesn't check out after sharding: timed out on 1 shard, but now takes 40 minutes per shard w/ 4 shards

https://hud.pytorch.org/hud/pytorch/pytorch/c9d12f6360e1af553046553ee57625fc7e036f3c/3?per_page=100&name_filter=aarch

Before sharding commit: 
https://hud.pytorch.org/pytorch/pytorch/commit/37a08b33bb86ad0fce3db6fc98f90c9e02d77cba

After sharding commit: 
https://hud.pytorch.org/pytorch/pytorch/commit/701ba5203fe68d55d655bd4d6c008be94cf34ea5

Looking through the logs:
No tests are missing or run multiple times after sharding
The 40min x 4 shards but 210+ min x 1 shard discrepancy is because test_transformers and some other tests take much longer on 1 machine than when split up

For example, on the before sharding commit, test_transformers 3/3 took 84124s but 1749s after sharding.  Randomly selected test from that file `test_scaled_dot_product_fused_attention_mask_vs_math_cpu_fused_kernel0_float16_batch_size_12_q_seq_len_1030_kv_seq_len_17_n_head_3_head_dim_8_mask_dim_4_bool_mask_True_train_True_casual_False_set_attn_mask_True_cpu_float16` took 3s before sharding and <1s after

I notice that shards generally only run 1 shard of test_transformers per machine after sharding, so one possibility is that parallelism makes these tests slower due to resource starvation or causes some sort of hang in a test

**Comment by huydhn** [comment]
cc @clee2000 to resolve this issue

**Comment by clee2000** [comment]
Closing as the issue has been resolved and making https://github.com/pytorch/pytorch/issues/147387 to track the AI of checking that sharding is correct
